strawberryfields/gbs/subgraph.py

```diff
@@ -183,193 +178,93 @@ def random_search(
-            customized callable function. Options include:
+    When growth is required, the algorithm examines all nodes from the remainder of the graph as
+    candidates and adds-in the single node with the highest degree relative to the rest of the
+    subgraph. This results in a graph that is one node larger,  and if growth is still required
+    the algorithm performs the above procedure again.
 
-            - ``"greedy-density"``: resize subgraphs by iteratively adding/removing nodes that
-              result in the densest subgraph among all candidates (default)
-            - ``"greedy-degree"``: resize subgraphs by iteratively adding/removing nodes that
-              have the highest/lowest degree.
-            - *callable*: accepting ``(subgraphs: list, graph: nx.Graph,
-              target: int, resize_options: dict)`` as arguments and returning a ``list``
-              containing the resized subgraphs as lists of nodes, see :func:`greedy_density` for
-              an example signature
+    When shrinking is required, the algorithm examines all nodes from within the subgraph as
+    candidates and removes the single node with lowest degree relative to the subgraph. In both
+    growth and shrink phases, ties for addition/removal with nodes of equal degree are settled by
+    uniform random choice.
 
     Args:
-        subgraphs (iterable[list[int]]): an iterable over subgraphs, where each subgraph is given
-            by a list of nodes
+        subgraph (list[int]): a subgraph specified by a list of nodes
         graph (nx.Graph): the input graph
-        target (int): the target number of nodes (must be larger than 1)
-        resize_options (dict[str, Any]): dictionary specifying options used during resizing;
-            defaults to :const:`RESIZE_DEFAULTS`
+        min_size (int): minimum size for subgraph to be resized to
+        max_size (int): maximum size for subgraph to be resized to
 
     Returns:
-        list[list[int]]: a list of resized subgraphs, where each subgraph is given by a list of
-        nodes
+        dict[int, list[int]]: a dictionary of different sizes with corresponding subgraph
     """
-    resize_options = {**RESIZE_DEFAULTS, **(resize_options or {})}
-
-    try:
-        target = int(target)
-    except TypeError:
-        raise TypeError("target must be an integer")
-
-    if not 2 < target < graph.number_of_nodes():
-        raise ValueError(
-            "target must be greater than two and less than the number of nodes in "
-            "the input graph"
-        )
+    nodes = graph.nodes()
+    subgraph = set(subgraph)
 
-    method = resize_options["method"]
+    if not subgraph.issubset(nodes):
+        raise ValueError("Input is not a valid subgraph")
+    if min_size < 1:
+        raise ValueError("min_size must be at least 1")
+    if max_size >= len(nodes):
+        raise ValueError("max_size must be less than number of nodes in graph")
```

Comment:
```suggestion
        raise ValueError("max_size cannot be larger than the number of nodes in graph")
```

Reply:
Cause it can be equal, right?

Reply:
Currently picking a `max_size` to be equal to the total number of nodes in the graph _is_ prevented in the code.

The motivation is that I didn't see much point in providing that last resize step because it's always going to be the same, i.e. the whole graph itself.

But we can change this - what do you think?